File: runtime/ruby/units/rt/i32_div_s.rb
# requires: rt/s32, rt/trap
def i32_div_s(a, b)
  sa = s32(a)
  sb = s32(b)
  trap("integer divide by zero") if sb == 0
  q = sa.abs / sb.abs
  q = -q if (sa < 0) ^ (sb < 0)
  trap("integer overflow") if q > 0x7fff_ffff
  q & M32
end

File: runtime/perl/units/rt/i32_div_s.pl
# requires: rt/s32, rt/trap
# `use integer` division is C division: truncating, exactly wasm's div_s.
# The INT_MIN / -1 overflow would raise SIGFPE in C, so trap it first.
sub i32_div_s {
    my $sa = Rt::s32($_[0]);
    my $sb = Rt::s32($_[1]);
    Rt::trap('integer divide by zero') if $sb == 0;
    Rt::trap('integer overflow') if $sa == -2147483648 && $sb == -1;
    return (do { use integer; $sa / $sb }) & 0xFFFFFFFF;
}
